File: implementations/typescript/src/ast.ts
export type NumberValue = number;

export interface ParseError {
  message: string;
  line: number;
  column: number;
}

export interface BaseNode {
  kind: string;
}

export interface StringNode extends BaseNode {
  kind: "string";
  value: string;
}

export interface NumberNode extends BaseNode {
  kind: "number";
  value: NumberValue;
}

export interface BooleanNode extends BaseNode {
  kind: "boolean";
  value: boolean;
}

export interface ArrayNode extends BaseNode {
  kind: "array";
  elements: AstNode[];
}

export interface BlockNode extends BaseNode {
  kind: "block";
  name: string;
  argument?: string;
  properties: Record<string, AstNode>;
  blocks: Record<string, BlockNode>;
  named_blocks?: BlockNode[];
}

export interface DocumentNode extends BaseNode {
  kind: "document";
  blocks: BlockNode[];
}

export type AstNode = StringNode | NumberNode | BooleanNode | ArrayNode | BlockNode | DocumentNode;

File: packages/kotlin/src/main/kotlin/io/rcl/Ast.kt
package io.rcl

sealed interface AstNode { val kind: String }

data class DocumentNode(
  override val kind: String = "document",
  val blocks: List<BlockNode>,
) : AstNode

data class BlockNode(
  override val kind: String = "block",
  val name: String,
  val argument: String? = null,
  val properties: Map<String, AstNode> = emptyMap(),
  val blocks: Map<String, BlockNode> = emptyMap(),
  val namedBlocks: List<BlockNode> = emptyList(),
) : AstNode

data class StringNode(override val kind: String = "string", val value: String) : AstNode

data class NumberNode(override val kind: String = "number", val value: Double) : AstNode

data class BooleanNode(override val kind: String = "boolean", val value: Boolean) : AstNode

data class ArrayNode(override val kind: String = "array", val elements: List<AstNode>) : AstNode

data class ParseError(override val message: String, val line: Int, val column: Int) : RuntimeException("$message at $line:$column")
